### PR TITLE
Allows to modify the start time and end time properties of the SegmentSpan object.

### PR DESCRIPTION
**Pull request by qq362220083**
Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

___
### New feature
- Allows to modify the start time and end time properties of the SegmentSpan object.

  In a specific scenario, I want to modify the start time and end time properties of the SegmentSpan object.
  For example, in the diagnostic information of `StackExchange.Redis`(SERedis), the start time and end time are determined by SERedis

### CODE REVIEW

File: src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs [hunk 1]
```diff
@@ -81,7 +81,16 @@ public void AddLog(params LogEvent[] events)
...
         public void Finish()
         {
-            EndTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            if (EndTime == default || EndTime < StartTime)
+            {
+                EndTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            }
+        }
+
+        public void SetTimes(DateTimeOffset startTime, DateTimeOffset endTime)
```
Comment (wu-sheng) [inline src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs:90]: Why and when would you need to change manually? This is not a typical feature for SkyWalking.

Comment (wu-sheng) [reply src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs:90]: > For example, in the diagnostic information of StackExchange.Redis(SERedis), the start time and end time are determined by SERedis

I have read this, but it is not clear.

Comment (qq362220083) [reply src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs:90]: The collection method of SERedis is to collect all commands and time objects under the current link at one time.

like this: https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/Profiling/ProfiledCommandEnumerable.cs

it contains the exact time of the beginning and end of the event, so I hope to modify the time of the SegmentSpan object.

Comment (wu-sheng) [reply src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs:90]: @liuhaoyang Your call.

Comment (qq362220083) [reply src/SkyApm.Abstractions/Tracing/Segments/SegmentSpan.cs:90]: I tried to implement the `ISegmentContextFactory` interface, but still cannot modify the SegmentContext and SegmentSpan object properties.

**Review by liuhaoyang — CHANGES_REQUESTED** [review]
My suggestion is that you add StartTime to the API of `ISegmentContextFactory` and modify the Finish function of `SegmentSpan` to pass in the EndtTime parameter.
```
public class SegmentSpan 
{
 ...
      public interface ISegmentContextFactory
    {
        SegmentContext CreateEntrySegment(string operationName, ICarrier carrier,long stratTime = default(...));
   ...
    }
  public void Finish(long endTime = default(...))
  {
    ...
  }
}
```

**Comment by qq362220083** [comment]
> My suggestion is that you add StartTime to the API of `ISegmentContextFactory` and modify the Finish function of `SegmentSpan` to pass in the EndtTime parameter.
> 
> ```
> public interface ISegmentContextFactory
> {
>    SegmentContext CreateEntrySegment(string operationName, ICarrier carrier,long stratTime = default(...));
>    ...
> }
> public class SegmentSpan 
> {
>  ...
>   public void Finish(long endTime = default(...))
>   {
>     ...
>   }
> }
> ```

Okay, I have modified and commit it as suggested.

**Review by liuhaoyang — APPROVED** [review]

**Pull request merged**